#14243 AsyncResponse to avoid timeout when restoring big snapshots.

dsilvam opened the pull request.

dsilvam committed:
#14243 AsyncResponse to avoid timeout when restoring big snapshots.

dsilvam commented on dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java:
```diff
@@ -302,18 +309,15 @@ public Response snapshotIndex(@Context HttpServletRequest request, @PathParam("p
+                    indexAPI.deleteRepository(ESIndexAPI.BACKUP_REPOSITORY, true);
+                    snapshotFile.delete();
+                }
+            };
```
this is just replacing the anonymous class with lambda

dsilvam commented on dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java:
```diff
@@ -343,31 +347,36 @@ public void write(OutputStream os) throws IOException, WebApplicationException {
-        }catch(SnapshotRestoreException exc){
-        	return this.responseUtil.getErrorResponse(request, Response.Status.INTERNAL_SERVER_ERROR, Locale.getDefault(), null, "snapshot.fail");
-        }catch(InterruptedException exi){
-        	return this.responseUtil.getErrorResponse(request, Response.Status.SERVICE_UNAVAILABLE, Locale.getDefault(), null, "snapshot.upload.failed");
```
non of these are thrown anymore

dsilvam commented on dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java:
```diff
@@ -343,31 +347,36 @@ public void write(OutputStream os) throws IOException, WebApplicationException {
+                    response = this.responseUtil.getErrorResponse(request, Response.Status.SERVICE_UNAVAILABLE, Locale.getDefault(), null, "snapshot.upload.failed");
+                }
+
+                asyncResponse.resume(response);
```
we resume the response after we are done (good or bad) with the restoring

acurionedotcms commented on dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java:
```diff
@@ -343,31 +347,36 @@ public void write(OutputStream os) throws IOException, WebApplicationException {
+                Response response = null;
+                try {
+                    if(this.indexAPI.uploadSnapshot(inputFile)){
+                        response = Response.ok(new MessageEntity("Snapshot restored successfully")).build();
```
Can this be externalized ?

nollymar commented on dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java:
```diff
@@ -343,31 +347,36 @@ public void write(OutputStream os) throws IOException, WebApplicationException {
+                        response = Response.ok(new MessageEntity("Snapshot restored successfully")).build();
+                    }
+                } catch (InterruptedException | ExecutionException | IOException e) {
+                    response = this.responseUtil.getErrorResponse(request, Response.Status.SERVICE_UNAVAILABLE, Locale.getDefault(), null, "snapshot.upload.failed");
```
Is it worth to log this exception?

dsilvam replied:
I think so. Will add it

nollymar commented on dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java:
```diff
@@ -343,31 +347,36 @@ public void write(OutputStream os) throws IOException, WebApplicationException {
 
         }catch (SecurityException sec) {
             return this.responseUtil.getErrorResponse(request, Response.Status.UNAUTHORIZED, Locale.getDefault(), null, "authentication-failed");
         }catch(IllegalArgumentException iar){
```
same comment here

dsilvam replied:
will add them also

dsilvam committed:
ESIndexAPI.java: Remove not needed variable
ResponseUtil.java: Include method for transforming exception into response. Add case for 401/bad credentials
ESIndexResource.java: Standarize responses and use new util
WorkflowResource.java: Extract useful logic for transforming exceptions into responses and update calls
SecurityException.java: Include message in parent exception

dsilvam committed:
ESIndexAPI.java: throw DotStateException
ExceptionUtil.java: include DotStateException in the badrequest exception list
ESIndexResource.java: use ResponseUtil.mapExceptionResponse
Language.properties: new entry for successful snapshot upload
